unified-runtime/source/common/cuda-hip/stream_queue.hpp

```diff
@@ -0,0 +1,348 @@
+
+/// Generic implementation of an out-of-order UR queue based on in-order
+/// backend 'stream' objects.
+///
+/// This class is specifically designed for the CUDA and HIP adapters.
+template <typename ST, int CS, int TS> struct stream_queue_t {
+  using native_type = ST;
+  static constexpr int DefaultNumComputeStreams = CS;
+  static constexpr int DefaultNumTransferStreams = TS;
+
+  std::vector<native_type> ComputeStreams;
+  std::vector<native_type> TransferStreams;
+  // Stream used for recording EvQueue, which holds information about when the
+  // command in question is enqueued on host, as opposed to started. It is
+  // created only if profiling is enabled - either for queue or per event.
+  native_type HostSubmitTimeStream{0};
+  // delay_compute_ keeps track of which streams have been recently reused and
+  // their next use should be delayed. If a stream has been recently reused it
+  // will be skipped the next time it would be selected round-robin style. When
+  // skipped, its delay flag is cleared.
+  std::vector<bool> DelayCompute;
+  // keep track of which streams have applied barrier
+  std::vector<bool> ComputeAppliedBarrier;
+  std::vector<bool> TransferAppliedBarrier;
+  ur_context_handle_t_ *Context;
+  ur_device_handle_t_ *Device;
+  std::atomic_uint32_t RefCount;
+  std::atomic_uint32_t EventCount;
+  std::atomic_uint32_t ComputeStreamIndex;
+  std::atomic_uint32_t TransferStreamIndex;
+  unsigned int NumComputeStreams;
+  unsigned int NumTransferStreams;
+  unsigned int LastSyncComputeStreams;
+  unsigned int LastSyncTransferStreams;
+  unsigned int Flags;
+  ur_queue_flags_t URFlags;
+  int Priority;
+  // When ComputeStreamSyncMutex and ComputeStreamMutex both need to be
+  // locked at the same time, ComputeStreamSyncMutex should be locked first
+  // to avoid deadlocks
+  std::mutex ComputeStreamSyncMutex;
+  std::mutex ComputeStreamMutex;
+  std::mutex TransferStreamMutex;
+  std::mutex BarrierMutex;
+  bool HasOwnership;
+
+  stream_queue_t(std::vector<native_type> &&ComputeStreams,
+                 std::vector<native_type> &&TransferStreams,
+                 ur_context_handle_t_ *Context, ur_device_handle_t_ *Device,
+                 unsigned int Flags, ur_queue_flags_t URFlags, int Priority,
+                 bool BackendOwns = true)
+      : ComputeStreams{std::move(ComputeStreams)},
+        TransferStreams{std::move(TransferStreams)},
+        DelayCompute(this->ComputeStreams.size(), false),
+        ComputeAppliedBarrier(this->ComputeStreams.size()),
+        TransferAppliedBarrier(this->TransferStreams.size()), Context{Context},
+        Device{Device}, RefCount{1}, EventCount{0}, ComputeStreamIndex{0},
+        TransferStreamIndex{0}, NumComputeStreams{0}, NumTransferStreams{0},
+        LastSyncComputeStreams{0}, LastSyncTransferStreams{0}, Flags(Flags),
+        URFlags(URFlags), Priority(Priority), HasOwnership{BackendOwns} {
+    urContextRetain(Context);
+    urDeviceRetain(Device);
```

Comment:
DeviceRetain/Release can be removed, it's a no-op unless we expect `Device` to be a subdevice (I have a ticket open to rename and better document those entry points)